### DISCUSSION TITLE
Keyboard control

### DISCUSSION THREAD

**Discussion by kytu282** [osmandapp/OsmAnd#11383] [Q&A] [answered]
Good morning,

I am developing a keyboard to be able to control Osmand without using the screen, since I use it in the car and it is far away.
I would like to know what keys (keycode) are used to interact with the map, I have managed to move with the cursor keys and return to the center (key C), but I cannot make the zoom work. 
II would also like to know if there is any other function that is integrated, such as saving the current point or things like that. 
II use the Android platform, but I would also like to know if it works the same in IOS. 

Greetingss and thank you very much for your attention

**Comment by vshcherb** [comment, marked as answer]
It is '+' sign, but we've just added '=' here https://github.com/osmandapp/OsmAnd/pull/11412.

**Reply by kytu282** [reply]
I have tried using the + and - sign, but it doesn't work

**Reply by vshcherb** [reply]
Did you select External input device 'Keyboard'?

**Reply by marcoslgm** [reply]
With External input device 'Keyboard' selected, only works "+" and "-" keys for the zoom, but doesn't work the arrows keys to move the map.
Without External input device 'Keyboard' selected, works the arrows keys to move the map, but doesn't work "+" and "-" keys.

**Reply by kytu282** [reply]
Yes i have the same isue, with "keyboard" active, only zoom out works, but arrow keys and zoom in doesn't, without "keyboard" arrow keys works, but zoom doesn't

**Reply by vshcherb** [reply]
Arrows were fixed 2 weeks ago https://github.com/osmandapp/OsmAnd/commit/74fcdc7ce1080562630b502988f7798f3f39f2dc (did you try latest night build)? 
We've also decided to add +/- with "no keyboard" option
https://github.com/osmandapp/OsmAnd/pull/11469/files

**Reply by kytu282** [reply]
everything works perfectly in the nightly version, thank you very much for your work, I suppose that these modifications will come out in the stable 4.0 version, right?

**Reply by Emile31** [reply]
Hello !Is there a place where I can find all the keyboard keys mapped ? thanks

**Comment by openfnord** [comment]
Cool, a keyboard.

I suggest to write an **API e.g. a generic socket for telnet or to write a web service to talk with osmand!**

This is more generic and the keyboard goes to another app.
This app could then use a usb2can dongle to talk to your cars CAN bus or one could hack a bmw i-bus /K-Line, LIN, (or analog voltage buttons to raspi-pi connected using wifi) interface to catch the old style steering wheel buttons of bmw e39, e38, old rangerover cars, (or some other fancy buttons in your 1980s Jaguar) which had some bmw shity 1st/2nd gen navigation.
So hack your own key board app and have osmand only to implement the socket / web service. Web service would be better compatible for modern hipster stuff (node.js :-) ).  Which is easy to use and every car hacker can roll her own keyboard/car key integration.
I have an openwrt router in my car to give wifi to my osmand tablet. (Next tablet will be connected using usb-c nic to router. No wifi needed here.)

Example use case=cancel the navigation and just display map and be silent as ine changed his mind and drives to somewhere else.

Telnet/socket
osmand.nav.cancel=1
Answer:
200

Like in http status codes
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses

Webservice would use
192.268.23.42:9999/osmand/nav?cancel=1
Answer:
200

Other use case:
One saw a speed cam and wants to report it to openstreetmap.org

She presses the big red button on top of the instrument clyster to be syre to be quick and capture the lication.

The keyboard controller app sends

socket:
osmand.openstreetmap.action.AddPoiAtLoc=speedcam

192.268.23.42:9999/osmand/openstreetmap/action?AddPoiAtLoc=speedcam

The socket could also be implemented as usb.serial device e.g. stm32 other microcontroller but the kernel module needs to have a midule liaded whic could be problrmatic.

So I would vote for jut implementing a restful webservice for the remote control (not only keyboard) api

Cheers

**Reply by Lee-Carre** [reply]
This would also make it scriptable.
